crates/hir-ty/src/infer/pat.rs

```diff
@@ -553,6 +581,63 @@ impl InferenceContext<'_> {
             | Pat::Expr(_) => false,
         }
     }
+
+    fn try_resolve_slice_ty_to_array_ty(
+        &mut self,
+        before: &[PatId],
+        suffix: &[PatId],
+        slice: &Option<PatId>,
+    ) -> Option<Ty> {
+        if !slice.is_none() {
+            return None;
+        }
+
+        let len = before.len() + suffix.len();
+        let size =
+            consteval::usize_const(self.db, Some(len as u128), self.owner.krate(self.db.upcast()));
+
+        let elem_ty = self.table.new_type_var();
+        let array_ty = TyKind::Array(elem_ty.clone(), size).intern(Interner);
+        Some(array_ty)
+    }
+
+    /// Determines whether we can infer the expected type in the slice pattern to be of type array.
+    /// This is only possible if we're in an irrefutable pattern. If we were to allow this in refutable
+    /// patterns we wouldn't e.g. report ambiguity in the following situation:
+    ///
+    /// ```ignore(rust)
+    /// struct Zeroes;
+    /// const ARR: [usize; 2] = [0; 2];
+    /// const ARR2: [usize; 2] = [2; 2];
+    ///
+    /// impl Into<&'static [usize; 2]> for Zeroes {
+    ///     fn into(self) -> &'static [usize; 2] {
+    ///            &ARR
+    ///     }
+    /// }
+    ///
+    /// impl Into<&'static [usize]> for Zeroes {
+    ///     fn into(self) -> &'static [usize] {
+    ///        &ARR2
+    ///     }
+    /// }
+    ///
+    /// fn main() {
+    ///     let &[a, b]: &[usize] = Zeroes.into() else {
+    ///       ..
+    ///     };
+    /// }
+    /// ```
+    ///
+    /// If we're in an irrefutable pattern we prefer the array impl candidate given that
+    /// the slice impl candidate would be be rejected anyway (if no ambiguity existed).
+    fn decl_allows_array_type_infer(&self, decl_ctxt: Option<DeclContext>) -> bool {
```

Comment:
I'd prefer the name of the original rustc PR, that shows the intent more clearly behind this